### PR TITLE
XY Chart: Add area and scatter chart styles

### PR DESCRIPTION
**Pull request by thefinaljob**
Added area and scatter chart styles through a dropdown menu

Contributes towards fixing #45

### CODE REVIEW

**Review by MatthewKhouzam — COMMENTED** [review]
Initial comment

https://www.freecodecamp.org/news/writing-good-commit-messages-a-practical-guide/

**Comment by MatthewKhouzam** [comment]
Initial hot take:

* Setting the default to scatter is not a great idea. offering scatter is.
* Using a dropdown may not be as interesting as using toggle buttons
* Scatters show latencies better 
image
* The UI freezes in Linux/Firefox with this patch. Not sure if it's related.
* You need to sign off your patches if you want them to pass DCO
* Stacked area makes sense for very few views, I would omit it from the 1st patch.
* Most importantly, this is really cool! 

Initial action items:
* Fix your commit message
* Make line the default
* Use toggles (icons can be <i class="far fa-chart-scatter"></i> https://fontawesome.com/v5.15/icons/chart-line?style=regular)
* Remove the stacked area as it implies a total which is not the case often
* Have a github issue associated to the commit message

I only looked at the functionality, not the code, so this can be considered a "sanity check"

**Comment by MatthewKhouzam** [comment]
Also funny bug, the more I resize the view and play with it, the smaller the dropdown gets.
image

**Comment by ebugden** [comment]
I recently learned that each data series in an xy chart specifies the chart type/style (line, scatter) that should be used to visualize it. I think this PR should be adjusted to automatically **display data based on the chart type specified in the series label** rather than offering a drop down menu currently proposed. @MatthewKhouzam @tahini thoughts?

**Comment by tahini** [comment]
> I recently learned that each data series in an xy chart specifies the chart type/style (line, scatter) that should be used to visualize it. I think this PR should be adjusted to automatically **display data based on the chart type specified in the series label** rather than offering a drop down like I initially suggested to @thefinaljob. Allowing all chart types implies that any option makes sense which isn't the case (ex. you wouldn't use a scatter plot to visualize continuous time-based data). @MatthewKhouzam @tahini thoughts?

Yes! That was indeed discussed offline with Nikolai. Note that the data type from the server comes with the series (for instance, I think the CPU usage has the total as a line and individual threads as area), not the whole chart, while iiuc, with chartjs, the series type will have to be for the whole chart. Would you default to the first one you get?

File: packages/react-components/src/components/xy-output-component.tsx
Comment (tahini) [inline packages/react-components/src/components/xy-output-component.tsx:442]: As discussed offline, if different types of chart require the data to be formatted differently, they should be encapsulated together in a single class so that there can be no mismatch. You could add separate widget classes for each series type (in react, it is suggested to multiply small stateless classes instead of big ones with ton of stuff in it). So you could have ScatterXy, LineXy, AreaXy, which would take the series as props and would format the data appropriately and render the proper line type. 

Here, you could definitely have a race condition between changing the chart type and zooming in with the old chart type and get the view to temporarily render wrong, until it is rendered right ;-)

But I'll let the maintainers of the extension decide if my comment is purely informative or if you should act on it

**Comment by thefinaljob** [comment]
Reopeing after a snafu

**Comment by thefinaljob** [comment]
This is a mess, this is what happens when you try to fix solder.... several times.

**Comment by thefinaljob** [comment]
OK looks like the submission is ready. To address @MatthewKhouzam's comments.

- After discussing with Erica, we are keeping the dropdown as is, using a toggle for each individual chart style would take up too much real estate. Also, having a toggle between a chart and table style would not be optimal given that a lot of the time there isn't a table view available
- We cannot have the chart style be chosen automatically yet as the API is sending a "null" value. The endpoint exists but it is not usable as of yet. I will issue a bug report regarding this.
- Line has been made default
- The UI freeze typically happens when too many data points are chosen, this could be fixed by allowing for a max number of data points to visualized
- After discussing with Erica, the stacked area chart will stay. I forgot the reasons though @ebugden can chime in here :) 

Thank you for your feedback Matthew!! This was my first serious PR and it was great to get a lot of good pointers in the right direction. 

Known bugs
- Border is not set to 0 when changing from line to stacked area, after talking to @tahini it seems that only dividing each of these chart styles into separate components would fix the issue as they would be destroyed and re-initialized upon a change in the chart style.

**Comment by MatthewKhouzam** [comment]
Hey, before getting into the review proper, please remember to update the commit message to follow the guidelines.

The reasoning being if and when something goes wrong, it's jarring to have a Hodge-podge of commit messages to go through, standardized ones will make things clearer in our time of duress. :)

**Comment by MatthewKhouzam** [comment]
Stacked area charts freeze on my computer. Could you profile it with the apt trace CPU usage and clicking on all?

**Comment by ebugden** [comment]
> Hey, before getting into the review proper, please remember to update the commit message to follow the guidelines.
> 
> The reasoning being if and when something goes wrong, it's jarring to have a Hodge-podge of commit messages to go through, standardized ones will make things clearer in our time of duress. :)

@thefinaljob the commit guidelines were recently added to the main readme in the [How to contribute code](https://github.com/theia-ide/theia-trace-extension#how-to-contribute-code) section.

**Review by MatthewKhouzam — CHANGES_REQUESTED** [review]
Please update the commit message and fix the performance issue.

**Comment by mirsky-work** [comment]
I just tested the functionality on Gitpod - very cool!
A nitpicking - right now the [X, Y] values' label is shown when selecting a scatter chart element. Can we show it on mouse hover?

image

**Comment by ebugden** [comment]
Exceptionally, I'm going to make edits to the title and description of this PR to correct style issues.

**Comment by ebugden** [comment]
> any news on this one.

@arfio could take a look at this if they have free time.

**Comment by bhufmann** [comment]
Closing this one after #595 was merged. Thanks for the effort it helped with the current implementation.

**Pull request closed without merging**
